plugins/inputs/tencentcloudcm/tencentcloudcm.go

```diff
@@ -0,0 +1,448 @@
+		  ## - QCE/REDIS
+		  ## - QCE/LB_PUBLIC
+		  ## - QCE/LB_PRIVATE
+		  # [[inputs.tencentcloudcm.accounts.namespaces.regions.instances]]
+		  # [[inputs.tencentcloudcm.accounts.namespaces.regions.instances.dimensions]]
+		  #   name = ""
+		  #   value = ""
+`
+}
+
+// Description implements telegraf.Input interface
+func (t *TencentCloudCM) Description() string {
+	return "Pull Metric Statistics from Tencent Cloud Cloud Monitor"
+}
+
+// Init is for setup, and validating config.
+func (t *TencentCloudCM) Init() error {
+
+	if t.Period <= 0 {
+		return fmt.Errorf("period is empty")
+	}
+
+	if len(t.Accounts) == 0 {
+		return errors.New("account is empty")
+	}
+
+	// create account credential
+	for i := range t.Accounts {
+
+		if t.Accounts[i].SecretID == "" {
+			return fmt.Errorf("secret_id is empty")
+		}
+		if t.Accounts[i].SecretKey == "" {
+			return fmt.Errorf("secret_key is empty")
+		}
+
+		t.Accounts[i].Crs = common.NewCredential(
+			t.Accounts[i].SecretID,
+			t.Accounts[i].SecretKey,
+		)
+		if t.Accounts[i].Name == "" {
+			t.Accounts[i].Name = fmt.Sprintf("%v", i)
+		}
+		// check if namespace supports auto discovery
+		for _, namespace := range t.Accounts[i].Namespaces {
+			if namespace.Name == "" {
+				return fmt.Errorf("namespace is empty")
+			}
+			for _, region := range namespace.Regions {
+				if region.RegionName == "" {
+					return fmt.Errorf("region is empty")
+				}
+				_, ok := Registry[namespace.Name]
+				if len(region.Instances) == 0 && !ok {
+					return fmt.Errorf(
+						"unsupported namespace %s for discovering instances, please specify instances and dimensions",
+						namespace.Name,
+					)
+				}
+			}
+		}
+	}
+
+	// start discovery
+	t.discoverTool = NewDiscoverTool(t.Log)
+	go t.discoverTool.Discover(t.Accounts, t.DiscoveryInterval, t.Endpoint)
+	t.discoverTool.DiscoverMetrics()
+	for i := range t.Accounts {
+		for j := range t.Accounts[i].Namespaces {
+			// use discovered metrics if not specified
+			if len(t.Accounts[i].Namespaces[j].Metrics) == 0 {
+				metrics, ok := t.discoverTool.DiscoveredMetrics[t.Accounts[i].Namespaces[j].Name]
+				if !ok {
+					errorInfo := fmt.Sprintf("unsupported namespace %s for discovering metrics, please specify metrics",
+						t.Accounts[i].Namespaces[j].Name)
+					t.Log.Error(errorInfo)
+					return errors.New(errorInfo)
+				}
+				t.Accounts[i].Namespaces[j].Metrics = metrics
+			}
+		}
+	}
+
+	t.client = &cloudmonitorClient{Accounts: t.Accounts}
+
+	return nil
+
+}
+
+func (t *TencentCloudCM) updateWindow(relativeTo time.Time) {
+	windowEnd := relativeTo.Add(-time.Duration(t.Delay))
+
+	t.windowStart = windowEnd.Add(-time.Duration(t.Period) * 2)
+
+	t.windowEnd = windowEnd
+}
+
+// GetMetricObjects gets all metric objects
+func (t *TencentCloudCM) GetMetricObjects() []MetricObject {
+
+	// metricObejcts holds all metrics with it's corresponding region,
+	// namespace, credential and instances(dimensions) information.
+	metricObjects := []MetricObject{}
```

Comment:
It's a matter of taste, but
```suggestion
	var metricObjects []MetricObject
```
is also fine.